dnf5-plugins/needs_restarting_plugin/needs_restarting.cpp

```diff
@@ -0,0 +1,306 @@
+
+    // First, ask systemd for the boot time. If systemd is available, this is
+    // the best option.
+    try {
+        std::unique_ptr<sdbus::IConnection> connection;
+        connection = sdbus::createSystemBusConnection();
+        auto proxy = sdbus::createProxy("org.freedesktop.systemd1", "/org/freedesktop/systemd1");
+
+        const uint64_t systemd_boot_time_us =
+            proxy->getProperty("UserspaceTimestamp").onInterface(SYSTEMD_MANAGER_INTERFACE);
+
+        const time_t systemd_boot_time = static_cast<long>(systemd_boot_time_us) / (1000L * 1000L);
+
+        if (systemd_boot_time != 0) {
+            return systemd_boot_time;
+        }
+    } catch (const sdbus::Error & ex) {
+        // Some D-Bus error, maybe we're inside a container.
+    }
+
+    // Otherwise, take the maximum of the st_mtime of /proc/1 and /proc/uptime.
+    time_t proc_1_boot_time = 0;
+    struct stat proc_1_stat = {};
+    if (stat("/proc/1", &proc_1_stat) == 0) {
+        proc_1_boot_time = proc_1_stat.st_mtime;
+    }
+
+    time_t uptime_boot_time = 0;
+    std::ifstream uptime_stream{"/proc/uptime"};
+    if (uptime_stream.is_open()) {
+        double uptime = 0;
+        uptime_stream >> uptime;
+        if (uptime > 0) {
+            uptime_boot_time = std::time(nullptr) - static_cast<time_t>(uptime);
+        }
+    }
+
+    return std::max(proc_1_boot_time, uptime_boot_time);
+}
+
+libdnf5::rpm::PackageSet recursive_dependencies(
+    const libdnf5::rpm::Package & package, libdnf5::rpm::PackageQuery & installed) {
+    libdnf5::rpm::PackageSet dependencies{package.get_base()};
+    dependencies.add(package);
+
+    std::vector<libdnf5::rpm::Package> stack;
+    stack.emplace_back(package);
+
+    while (!stack.empty()) {
+        const auto & current = stack.back();
+        stack.pop_back();
+
+        libdnf5::rpm::PackageQuery query{installed};
+        query.filter_provides(current.get_requires());
+
+        for (const auto & dependency : query) {
+            if (!dependencies.contains(dependency)) {
+                stack.emplace_back(dependency);
+            }
+        }
+
+        dependencies |= query;
+    }
+
+    return dependencies;
+}
+
+void NeedsRestartingCommand::system_needs_restarting(Context & ctx) {
+    const auto boot_time = get_boot_time();
+
+    libdnf5::rpm::PackageQuery base_query{ctx.base};
+
+    libdnf5::rpm::PackageQuery installed{base_query};
+    installed.filter_installed();
+
+    libdnf5::rpm::PackageQuery reboot_suggested{installed};
+    reboot_suggested.filter_reboot_suggested();
+
+    std::vector<libdnf5::rpm::Package> need_reboot = {};
+    for (const auto & pkg : reboot_suggested) {
+        if (pkg.get_install_time() > static_cast<unsigned long long>(boot_time)) {
+            need_reboot.push_back(pkg);
+        }
+    }
+
+    if (need_reboot.empty()) {
+        std::cout << "No core libraries or services have been updated since boot-up.\n"
```

Comment:
Minor: would be better to use `std::endl` instead of `\n` on the output (more occurrences in this file).